### ISSUE TITLE
Index created on 2.x causing "unsupported_operation_exception" in 5.x when used for sorting

### ISSUE THREAD

**Issue by jakommo** [elastic/elasticsearch#26162]
**Elasticsearch version** (`bin/elasticsearch --version`): 5.5.0

**Plugins installed**: []

**JVM version** (`java -version`): Java 8 u131

**OS version** (`uname -a` if on a Unix-like system): Ubuntu 16.04

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

## Create index in 2.x (I used 2.3.4)

```
DELETE test1
DELETE test2

PUT test1
{
  "mappings": {
    "_default_": {
      "properties": {
        "testfield": {
          "type": "string", "index": "not_analyzed"
        }
      }
    }
  }
}

PUT test2
{
  "mappings": {
    "_default_": {
      "properties": {
        "testfield": {
          "type": "string", "index": "not_analyzed"
        }
      }
    }
  }
}

POST test1/test
{
  "testfield": "my test"
}

# Don't index anything into test2

POST test*/_search
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "testfield": {
        "order": "asc",
        "unmapped_type": "string"
      }
    }
  ]
}
```

## Upgrade to 5.x (I used 5.5.0)

The query above, now results in: (Also if I change to "unmapped_type": "keyword")

```
{
  "took": 19,
  "timed_out": false,
  "_shards": {
    "total": 10,
    "successful": 5,
    "failed": 5,
    "failures": [
      {
        "shard": 0,
        "index": "test2",
        "node": "JBHJ1HPYQbKMOxJ1hSotrw",
        "reason": {
          "type": "unsupported_operation_exception",
          "reason": null
        }
      }
    ]
  },
  "hits": {
    "total": 1,
    "max_score": null,
    "hits": [
      {
        "_index": "test1",
        "_type": "test",
        "_id": "AV3Qh0G3rdu9hzgAfZn2",
        "_score": null,
        "_source": {
          "testfield": "my test"
        },
        "sort": [
          "my test"
        ]
      }
    ]
  }
}
```
and is producing this stack trace:

```
[2017-08-11T11:13:16,153][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][0], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=OCMZA_bHTA-cdS6nEHIBgA]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
[2017-08-11T11:13:16,155][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][2], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=kkIHhfYIRf6vXlBlebIpgg]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
[2017-08-11T11:13:16,157][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][3], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=GyLowxvgSeO8ZabUK13b4g]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
[2017-08-11T11:13:16,161][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][1], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=WEeR_WDhSW-Uxkys3ChfQg]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
[2017-08-11T11:13:16,167][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][4], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=ka-M_4BJStSmPorrvv2_Cw]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]

```

As soon as I index a doc in `test2`, the error goes away and the query is working:

```
POST test2/test
{
  "testfield": "my test"
}

POST test*/_search
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "testfield": {
        "order": "asc",
        "unmapped_type": "keyword"
      }
    }
  ]
}
``` 
now results in:

```
{
  "took": 8,
  "timed_out": false,
  "_shards": {
    "total": 10,
    "successful": 10,
    "failed": 0
  },
  "hits": {
    "total": 2,
    "max_score": null,
    "hits": [
      {
        "_index": "test1",
        "_type": "test",
        "_id": "AV3Qh0G3rdu9hzgAfZn2",
        "_score": null,
        "_source": {
          "testfield": "my test"
        },
        "sort": [
          "my test"
        ]
      },
      {
        "_index": "test2",
        "_type": "test",
        "_id": "AV3Qku9k6j_T2EVH4QtG",
        "_score": null,
        "_source": {
          "testfield": "my test"
        },
        "sort": [
          "my test"
        ]
      }
    ]
  }
}
```

**Comment by markharwood** [comment]
Note: The point of referencing 2.3.4 in this issue is not to show "this used to work, now it doesn't".
It seems the bug only manifests itself on indexes that have undergone a 2.3.4 - > 5.x upgrade step. I tried setting up these indices from fresh on 5.5 without performing the 2.3.4 upgrade and queries were OK. 
I suspect it's quite an edge case with the combination of default mappings, upgrades and querying empty indexes to consider - is this affecting any production systems?

**Comment by jakommo** [comment]
Thanks @markharwood 

> Note: The point of referencing 2.3.4 in this issue is not to show "this used to work, now it doesn't".
It seems the bug only manifests itself on indexes that have undergone a 2.3.4 - > 5.x upgrade step. I tried setting up these indices from fresh on 5.5 without performing the 2.3.4 upgrade and queries were OK.

Yes, sorry for not making this clear. It only affects upgrades. If you create an index in 5.x, it will convert the string/not_analyzed to type keyword in the background. 

> I suspect it's quite an edge case with the combination of default mappings, upgrades and querying empty indexes to consider - is this affecting any production systems?

It was affecting my dev env, but it could also have happened on the prod system.
Yes, it's quite an edge case, but I'm not sure what the root cause is and if it could have other effects as well.

**Comment by rodgermoore** [comment]
Ran into the same problem. Workaround in https://github.com/elastic/kibana/issues/13950 fixed my Kibana index. Thanks again @jakommo .

**Comment by tomsommer** [comment]
This will be a bug that keeps coming back, until 5.6.1 is released. Any chance of doing this ASAP?

**Comment by tomsommer** [comment]
5.6.1 is out, so I guess this is a non-issue now?

**Comment by jasontedor** [comment]
Closed by #26602
